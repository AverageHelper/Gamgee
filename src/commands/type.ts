import type { Command } from "./Command";
import { deleteMessage } from "../actions/messages/deleteMessage";

const type: Command = {
  name: "t",
  description: "Start typing. :wink:",
  async execute({ message, client, logger }) {
    const channel = message.channel;

    logger.debug(
      `${client.user?.username.concat(" is") ?? "I am"} typing in channel ${channel.id}...`
    );
    await deleteMessage(message, "Spam: Users don't need to see this command run");

    setTimeout(() => {
      channel.stopTyping(true);
    }, 5000);

    await channel.startTyping();
    logger.debug(`Finished typing in channel ${channel.id}`);
  }
};

export default type;
