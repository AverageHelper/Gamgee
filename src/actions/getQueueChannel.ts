import type Discord from "discord.js";
import type { CommandContext } from "../commands/index";
import { getConfigQueueChannel } from "../actions/config/getConfigValue";
import { useLogger } from "../logger";

const logger = useLogger();

/**
 * Retrieves the configured queue channel. Returns `null` if none has been set up yet.
 *
 * This action may sent error messages to the channel using the provided `context`.
 *
 * @param context The context of the command that needs the queue channel.
 * @returns
 */
export default async function getQueueChannel(
  context: CommandContext
): Promise<Discord.TextChannel | null> {
  const { message, client, storage } = context;

  const queueChannelId = await getConfigQueueChannel(storage);
  if (!queueChannelId) {
    return null;
  }

  let queueChannel: Discord.TextChannel;
  try {
    queueChannel = (await client.channels.fetch(queueChannelId)) as Discord.TextChannel;
  } catch (error) {
    logger.error("Failed to fetch queue channel:", error);
    await message.channel.send(
      "The configured channel doesn't exist. Have an administrator set the queue back up."
    );
    return null;
  }

  if (!queueChannel.isText()) {
    logger.error("The configured channel is not a text channel.");
    await message.channel.send(
      "The configured channel is not a text channel. Have an administrator set up the queue again."
    );
    return null;
  }

  return queueChannel;
}
