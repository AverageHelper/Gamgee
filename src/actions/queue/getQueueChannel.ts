import type { CommandContext } from "../../commands/index.js";
import type Discord from "discord.js";
import { getQueueChannelId } from "../../useGuildStorage.js";
import { richErrorMessage } from "../../helpers/richErrorMessage.js";
import { useLogger } from "../../logger.js";

const logger = useLogger();

async function getQueueChannelFromCommand(
	context: CommandContext
): Promise<Discord.TextChannel | null> {
	if (!context.guild) return null;

	const queueChannelId = await getQueueChannelId(context.guild);
	if (queueChannelId === null || !queueChannelId) return null;

	let queueChannel: Discord.Channel | null;
	try {
		queueChannel = await context.client.channels.fetch(queueChannelId);
	} catch (error) {
		logger.error(richErrorMessage("Failed to fetch queue channel.", error));
		await context.reply(
			"The configured channel doesn't exist. Have an administrator set the queue back up."
		);
		return null;
	}

	if (!queueChannel) return null;

	if (!queueChannel.isText()) {
		logger.error("The configured channel is not a text channel.");
		await context.reply(
			"The configured channel is not a text channel. Have an administrator set up the queue again."
		);
		return null;
	}

	return queueChannel as Discord.TextChannel;
}

async function getQueueChannelFromGuild(guild: Discord.Guild): Promise<Discord.TextChannel | null> {
	const queueChannelId = await getQueueChannelId(guild);
	if (queueChannelId === null || !queueChannelId) {
		return null;
	}

	let queueChannel: Discord.TextChannel;
	try {
		queueChannel = (await guild.client.channels.fetch(queueChannelId)) as Discord.TextChannel;
	} catch (error) {
		logger.error(richErrorMessage("Failed to fetch queue channel.", error));
		return null;
	}

	if (!queueChannel.isText()) {
		logger.error("The configured channel is not a text channel.");
		return null;
	}

	return queueChannel;
}

/**
 * Retrieves the configured queue channel. Returns `null` if none has been set up yet.
 *
 * This action may send error messages to the message's channel.
 *
 * @param source The guild or command invocation from which to derive a queue channel.
 *
 * @returns the guild's queue channel, or `null` if it has none.
 */
export async function getQueueChannel(
	source: CommandContext | Discord.Guild | null
): Promise<Discord.TextChannel | null> {
	if (!source) return null;
	if ("type" in source) return getQueueChannelFromCommand(source);
	return getQueueChannelFromGuild(source);
}
