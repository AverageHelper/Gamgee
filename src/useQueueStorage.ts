import type { Snowflake, TextChannel } from "discord.js";
import { DEFAULT_MESSAGE_COMMAND_PREFIX } from "./constants/database.js";
import { useLogger } from "./logger.js";
import { useRepository } from "./database/useDatabase.js";
import type {
	QueueConfig as _QueueConfig,
	QueueConfigToBlacklistedUsers,
	QueueEntry as _QueueEntry,
	User
} from "@prisma/client";

const logger = useLogger();

export type QueueEntry = _QueueEntry & { haveCalledNowPlaying: Array<User> };
export type UnsentQueueEntry = Omit<
	_QueueEntry,
	"queueMessageId" | "isDone" | "channelId" | "guildId" | "sentAt" | "haveCalledNowPlaying"
>;

// TODO: Break these into separate files where appropriate
// TODO: Make clear that these functions do not touch Discord's API, only local storage
// TODO: Only retain user-provided data for at most 90 days

// ** Queue Config **

export type QueueConfig = _QueueConfig & {
	blacklistedUsers: Array<User>;
};

/**
 * Retrieves the queue's configuration settings.
 *
 * @param queueChannel The channel that identifies the request queue.
 *
 * @returns a promise that resolves with the queue config for the channel
 * or a default one if none has been set yet.
 */
export async function getQueueConfig(queueChannel: TextChannel): Promise<QueueConfig> {
	const extantConfig = await useRepository("queueConfig", queueConfigs =>
		queueConfigs.findUnique({
			where: { channelId: queueChannel.id },
			include: {
				blacklistedUsers: {
					include: { user: true }
				}
			}
		})
	);
	return {
		channelId: queueChannel.id,
		entryDurationSeconds: extantConfig?.entryDurationSeconds ?? null,
		entryDurationMinSeconds: extantConfig?.entryDurationMinSeconds ?? null,
		queueDurationSeconds: extantConfig?.queueDurationSeconds ?? null,
		cooldownSeconds: extantConfig?.cooldownSeconds ?? null,
		submissionMaxQuantity: extantConfig?.submissionMaxQuantity ?? null,
		blacklistedUsers: extantConfig?.blacklistedUsers.map(u => u.user) ?? []
	};
}

/**
 * Updates the provided properties of a queue's configuration settings.
 *
 * @param config Properties of the queue config to overwrite the current data.
 * @param queueChannel The channel that identifies the request queue.
 */
export async function updateQueueConfig(
	config: Partial<QueueConfig>,
	queueChannel: TextChannel
): Promise<void> {
	if (Object.keys(config).length === 0) return; // nothing to store

	// Compose updates
	const update: Partial<_QueueConfig> & {
		blacklistedUsers?: {
			set: Array<{ queueConfigsChannelId_userId: QueueConfigToBlacklistedUsers }>;
		};
	} = {
		entryDurationSeconds: config.entryDurationSeconds,
		entryDurationMinSeconds: config.entryDurationMinSeconds,
		queueDurationSeconds: config.queueDurationSeconds,
		cooldownSeconds: config.cooldownSeconds,
		submissionMaxQuantity: config.submissionMaxQuantity
	};

	if (config.blacklistedUsers !== undefined) {
		update.blacklistedUsers = {
			set: config.blacklistedUsers.map(user => ({
				queueConfigsChannelId_userId: { queueConfigsChannelId: queueChannel.id, userId: user.id }
			}))
		};
	}

	// Update or create the config
	await useRepository("queueConfig", configs =>
		configs.upsert({
			where: { channelId: queueChannel.id },
			update,

			create: {
				channelId: queueChannel.id,
				...update
			}
		})
	);
}

// ** Write Song Entries **

/**
 * Adds the queue entry to the database.
 *
 * @param entry Properties of the new request entity.
 * @param queueChannel The channel that identifies the request queue.
 *
 * @returns a promise that resolves with the new queue entry
 */
export async function createEntry(
	entry: Omit<_QueueEntry, "channelId" | "guildId">,
	queueChannel: TextChannel
): Promise<QueueEntry> {
	// FIXME: These could be done all in one go if we used table relations properly

	// Make sure the guild config is in there
	await useRepository("guild", guilds =>
		guilds.upsert({
			where: { id: queueChannel.guildId },
			update: {},

			create: {
				currentQueue: queueChannel.id,
				id: queueChannel.guildId,
				isQueueOpen: false,
				messageCommandPrefix: DEFAULT_MESSAGE_COMMAND_PREFIX
			}
		})
	);

	// Make sure the channel is in there
	await useRepository("channel", channels =>
		channels.upsert({
			where: { id: queueChannel.id },
			update: {},

			create: {
				guildId: queueChannel.guildId,
				id: queueChannel.id
			}
		})
	);

	// Make sure we have at least the default queue config
	await useRepository("queueConfig", queueConfigs =>
		queueConfigs.upsert({
			where: { channelId: queueChannel.id },
			update: {},

			create: {
				channelId: queueChannel.id,
				entryDurationSeconds: null,
				cooldownSeconds: null,
				submissionMaxQuantity: null,
				queueDurationSeconds: null,
				entryDurationMinSeconds: null,
				blacklistedUsers: undefined
			}
		})
	);

	const newEntry = {
		channelId: queueChannel.id,
		guildId: queueChannel.guildId,
		isDone: entry.isDone,
		queueMessageId: entry.queueMessageId,
		seconds: entry.seconds,
		senderId: entry.senderId,
		sentAt: entry.sentAt,
		url: entry.url
	};

	// Add the entry, or update the one we have
	return await useRepository("queueEntry", queueEntries =>
		queueEntries.upsert({
			where: {
				queueMessageId: entry.queueMessageId
			},
			update: newEntry,

			create: newEntry,

			include: { haveCalledNowPlaying: true }
		})
	);
}

/**
 * Removes the queue entry from the database.
 *
 * @param queueMessageId The ID of the message that identifies the entry in the queue channel.
 */
export async function removeEntryFromMessage(queueMessageId: Snowflake): Promise<void> {
	await useRepository("queueEntry", queueEntries =>
		queueEntries.delete({
			where: { queueMessageId }
		})
	);
}

// ** Read Song Entries **

/**
 * Fetches an entry with the given message ID.
 *
 * @param queueMessageId The ID of the message that identifies the entry in the queue channel.
 *
 * @returns a promise that resolves with the matching queue entry
 * or `null` if no such entry exists
 */
export async function fetchEntryFromMessage(queueMessageId: Snowflake): Promise<QueueEntry | null> {
	return await useRepository("queueEntry", queueEntries =>
		queueEntries.findUnique({
			where: { queueMessageId },
			include: { haveCalledNowPlaying: true }
		})
	);
}

/**
 * Fetches all entries in the queue, in order of appearance.
 *
 * @param queueChannel The channel that identifies the request queue.
 * @returns a promise that resolves with the queue's entries,
 * in the order in which they were added.
 */
export async function fetchAllEntries(queueChannel: TextChannel): Promise<Array<QueueEntry>> {
	return await useRepository("queueEntry", queueEntries =>
		queueEntries.findMany({
			where: {
				channelId: queueChannel.id,
				guildId: queueChannel.guild.id
			},
			orderBy: { sentAt: "asc" },
			include: { haveCalledNowPlaying: true }
		})
	);
}

/**
 * Fetches the number of entries in the queue.
 *
 * @param queueChannel The channel that identifies the request queue.
 * @returns a promise that resolves with the number of entries in the queue.
 */
export async function countAllEntries(queueChannel: TextChannel): Promise<number> {
	return await useRepository("queueEntry", queueEntries =>
		queueEntries.count({
			where: {
				channelId: queueChannel.id,
				guildId: queueChannel.guildId
			}
		})
	);
}

/**
 * Fetches all entries by the given user.
 *
 * @param senderId The ID of the user who submitted entries.
 * @param queueChannel The channel that identifies the request queue.
 * @returns a promise that resolves with the user's entries,
 * in the order in which they were added.
 */
export async function fetchAllEntriesFrom(
	senderId: string,
	queueChannel: TextChannel
): Promise<Array<QueueEntry>> {
	return await useRepository("queueEntry", queueEntries =>
		queueEntries.findMany({
			where: {
				channelId: queueChannel.id,
				guildId: queueChannel.guild.id,
				senderId
			},
			orderBy: { sentAt: "asc" },
			include: { haveCalledNowPlaying: true }
		})
	);
}

/**
 * Fetches the lastest entry by the given user.
 *
 * @param senderId The ID of the user who submitted entries.
 * @param queueChannel The channel that identifies the request queue.
 * @returns a promise that resolves with the user's latest entry
 * or `null` if the user has no associated entries.
 */
export async function fetchLatestEntryFrom(
	senderId: string,
	queueChannel: TextChannel
): Promise<QueueEntry | null> {
	return await useRepository("queueEntry", queueEntries =>
		queueEntries.findFirst({
			where: {
				channelId: queueChannel.id,
				guildId: queueChannel.guild.id,
				senderId
			},
			orderBy: { sentAt: "desc" },
			include: { haveCalledNowPlaying: true }
		})
	);
}

/**
 * Fetches the number of entries from the given user in the queue.
 *
 * @param senderId The ID of the user who submitted entries.
 * @param queueChannel The channel that identifies the request queue.
 *
 * @returns a promise that resolves with the number of entries in
 * the queue associated with the user.
 */
export async function countAllEntriesFrom(
	senderId: string,
	queueChannel: TextChannel
): Promise<number> {
	return await useRepository("queueEntry", queueEntries =>
		queueEntries.count({
			where: {
				channelId: queueChannel.id,
				guildId: queueChannel.guild.id,
				senderId
			}
		})
	);
}

/** Returns the average entry duration of the submissions of the user with the provided ID. */
export async function averageSubmissionPlaytimeForUser(
	userId: Snowflake,
	queueChannel: TextChannel
): Promise<number> {
	const entries = await fetchAllEntriesFrom(userId, queueChannel);
	let average = 0;

	entries.forEach(entry => {
		average += entry.seconds;
	});
	average /= entries.length;

	return average;
}

/**
 * Sets the entry's "done" value.
 *
 * @param isDone Whether the entry should be marked "done"
 * @param queueMessageId The ID of the message that identifies the entry in the queue channel.
 */
export async function markEntryDone(isDone: boolean, queueMessageId: Snowflake): Promise<void> {
	logger.debug(`Marking entry ${queueMessageId} as ${isDone ? "" : "not "}done`);
	await useRepository("queueEntry", queueEntries =>
		queueEntries.update({
			where: { queueMessageId },
			data: { isDone }
		})
	);
}

/**
 * Deletes all request entries for this queue from the database.
 * @param queueChannel The channel that identifies the request queue.
 */
export async function clearEntries(queueChannel: TextChannel): Promise<void> {
	await useRepository("queueEntry", queueEntries =>
		queueEntries.deleteMany({
			where: {
				channelId: queueChannel.id,
				guildId: queueChannel.guild.id
			}
		})
	);
}

// ** Now-playing Invocations **

/**
 * Retrieves the number of users that invoked `/nowplaying` or its variants for an entry,
 * excluding the requesting user if they also invoked that command.
 *
 * @param queueMessageId The ID of the message that identifies the entry in the queue channel.
 * @param queueChannel The channel that identifies the request queue.
 * @returns a promise that resolves with the number of unique `/nowplaying` invocations
 * for the entry.
 */
export async function getLikeCount(
	queueMessageId: Snowflake,
	queueChannel: TextChannel
): Promise<number> {
	const entry = await useRepository("queueEntry", queueEntries =>
		queueEntries.findFirst({
			where: {
				channelId: queueChannel.id,
				guildId: queueChannel.guild.id,
				queueMessageId
			},
			select: { haveCalledNowPlaying: true }
		})
	);
	return entry?.haveCalledNowPlaying.length ?? Number.NaN;
}

/**
 * Increments the count of unique `/nowplaying` invocations for the entry.
 * If the user submitted this entry, they are not counted.
 * If the user has already invoked `/nowplaying` for this entry, they are not counted.
 *
 * @param userId The ID of the user who invoked `/nowplaying`.
 * @param queueMessageId The ID of the message that identifies the entry in the queue channel.
 * @param queueChannel The channel that identifies the request queue.
 */
export async function addToHaveCalledNowPlaying(
	userId: Snowflake,
	queueMessageId: Snowflake,
	queueChannel: TextChannel
): Promise<void> {
	logger.debug(`Adding ${userId} to haveCalledNowPlaying for ${queueMessageId}`);
	const entry = await useRepository("queueEntry", queueEntries =>
		queueEntries.findFirst({
			where: {
				channelId: queueChannel.id,
				guildId: queueChannel.guild.id,
				queueMessageId
			},
			select: { senderId: true, haveCalledNowPlaying: true }
		})
	);
	if (!entry) {
		logger.debug(`Could not find entry '${queueMessageId}' in the database!`);
		return;
	}
	if (userId === entry.senderId) {
		logger.debug("User calling is this request's sender, skipping");
		return;
	}
	if (entry?.haveCalledNowPlaying.map(u => u.id).includes(userId)) {
		logger.debug("User in haveCalledNowPlaying already, skipping...");
		return;
	}
	await useRepository("queueEntry", queueEntries =>
		queueEntries.update({
			where: { queueMessageId },
			data: {
				haveCalledNowPlaying: {
					connect: {
						id: userId
					}
				}
			}
		})
	);
	logger.debug("User added to haveCalledNowPlaying");
}

// ** User Blacklist **

/**
 * Adds the user to the queue's blacklist. That user will not be able to submit song requests.
 *
 * @param userId The ID of the user to blacklist.
 * @param queueChannel The channel that identifies the request queue.
 */
export async function blacklistUser(userId: Snowflake, queueChannel: TextChannel): Promise<void> {
	const blacklistedUsers = {
		connectOrCreate: {
			where: {
				queueConfigsChannelId_userId: {
					queueConfigsChannelId: queueChannel.id,
					userId
				}
			},
			create: { userId } // the User to create if none were found to match the above reference
		}
	};

	await useRepository("queueConfig", queueConfigs =>
		queueConfigs.upsert({
			// If a config is found, update it:
			where: { channelId: queueChannel.id },
			update: { blacklistedUsers },

			// If the queue config isn't found, create it:
			create: {
				channelId: queueChannel.id,
				entryDurationSeconds: null,
				entryDurationMinSeconds: null,
				queueDurationSeconds: null,
				cooldownSeconds: null,
				submissionMaxQuantity: null,
				blacklistedUsers
			}
		})
	);
}

/**
 * Removes the user from the queue's blacklist.
 *
 * @param userId The ID of the user to whitelist.
 * @param queueChannel The channel that identifies the request queue.
 */
export async function whitelistUser(userId: Snowflake, queueChannel: TextChannel): Promise<void> {
	await useRepository("queueConfig", queueConfigs =>
		queueConfigs.update({
			where: { channelId: queueChannel.id },
			data: {
				blacklistedUsers: {
					disconnect: {
						queueConfigsChannelId_userId: {
							queueConfigsChannelId: queueChannel.id,
							userId
						}
					}
				}
			}
		})
	);
}
