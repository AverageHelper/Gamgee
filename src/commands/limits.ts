import type { Command } from "./Command.js";
import { allLimits } from "./queue/limit.js";
import { assertUnreachable } from "../helpers/assertUnreachable.js";
import { code } from "../helpers/composeStrings.js";
import { durationString } from "../helpers/durationString.js";
import { EmbedBuilder } from "discord.js";
import { getQueueChannel } from "../actions/queue/getQueueChannel.js";
import { getStoredQueueConfig } from "../useQueueStorage.js";
import { localizations, t, ti } from "../i18n.js";

export const limits: Command = {
	name: "limits",
	nameLocalizations: localizations("commands.limits.name"),
	description: "Display the song queue's submission limits.",
	descriptionLocalizations: localizations("commands.limits.description"),
	requiresGuild: true,
	async execute({ type, guild, guildLocale, userLocale, reply }) {
		const queueChannel = await getQueueChannel(guild);
		if (!queueChannel) {
			return await reply(t("common.queue.not-set-up", guildLocale));
		}

		const config = await getStoredQueueConfig(queueChannel);

		// We call `reply` with `ephemeral: true`, which has different behavior between
		// slash-command interactions and message commands.
		const locale = type === "interaction" ? userLocale : guildLocale;

		const { cooldown: cooldownCommand } = await import("./cooldown.js");
		const cooldownName: string = cooldownCommand.nameLocalizations
			? cooldownCommand.nameLocalizations[locale] ?? cooldownCommand.name
			: cooldownCommand.name;

		// Read out the existing limits
		const embed = new EmbedBuilder()
			.setTitle(t("commands.limits.responses.title", locale))
			.setDescription(
				ti(
					"commands.limits.responses.use-cooldown-cmd",
					{ cooldown: code(`/${cooldownName}`) },
					locale
				)
			);

		allLimits(locale).forEach(key => {
			let value: string;
			switch (key.value) {
				case "cooldown":
					if (config.cooldownSeconds !== null && config.cooldownSeconds > 0) {
						value = durationString(locale, config.cooldownSeconds);
					} else {
						value = t("commands.limits.responses.none", locale);
					}
					break;
				case "entry-duration-max":
					if (config.entryDurationMaxSeconds !== null && config.entryDurationMaxSeconds > 0) {
						value = durationString(locale, config.entryDurationMaxSeconds);
					} else {
						value = t("commands.limits.responses.infinite", locale);
					}
					break;
				case "entry-duration-min":
					if (config.entryDurationMinSeconds !== null && config.entryDurationMinSeconds > 0) {
						value = durationString(locale, config.entryDurationMinSeconds);
					} else {
						value = "0";
					}
					break;
				case "queue-duration":
					if (config.queueDurationSeconds !== null && config.queueDurationSeconds > 0) {
						value = durationString(locale, config.queueDurationSeconds);
					} else {
						value = t("commands.limits.responses.infinite", locale);
					}
					break;
				case "count":
					if (config.submissionMaxQuantity !== null && config.submissionMaxQuantity > 0) {
						value = `${config.submissionMaxQuantity}`;
					} else {
						value = t("commands.limits.responses.infinite", locale);
					}
					break;
				default:
					assertUnreachable(key.value);
			}

			embed.addFields({ name: `${key.name}:\t${value}`, value: key.description });
		});

		await reply({ embeds: [embed], ephemeral: true });
	}
};
