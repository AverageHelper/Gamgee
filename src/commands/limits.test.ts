import type { Mock } from "vitest";
import { beforeEach, describe, expect, test, vi } from "vitest";

vi.mock("../useQueueStorage.js");
vi.mock("../actions/queue/getQueueChannel.js");
vi.mock("../actions/queue/useQueue.js");

import { getStoredQueueConfig } from "../useQueueStorage.js";
const mockGetStoredQueueConfig = getStoredQueueConfig as Mock<
	Parameters<typeof getStoredQueueConfig>,
	ReturnType<typeof getStoredQueueConfig>
>;

import { getQueueChannel } from "../actions/queue/getQueueChannel.js";
const mockGetQueueChannel = getQueueChannel as Mock<
	Parameters<typeof getQueueChannel>,
	ReturnType<typeof getQueueChannel>
>;

import type { GuildedCommandContext } from "./CommandContext.js";
import type { TextChannel } from "discord.js";
import { limits } from "./limits.js";

const mockReply = vi.fn().mockResolvedValue(undefined);

describe("Get Queue Limits", () => {
	let context: GuildedCommandContext;

	beforeEach(() => {
		context = {
			guild: "the-guild",
			user: { id: "the-user" },
			reply: mockReply
		} as unknown as GuildedCommandContext;

		mockGetQueueChannel.mockResolvedValue({
			id: "queue-channel"
		} as unknown as TextChannel);
		mockGetStoredQueueConfig.mockResolvedValue({
			cooldownSeconds: null,
			entryDurationMaxSeconds: null,
			submissionMaxQuantity: null,
			blacklistedUsers: [],
			channelId: "",
			queueDurationSeconds: null,
			entryDurationMinSeconds: null
		});
	});

	test("cannot show statistics on a queue that does not exist", async () => {
		mockGetQueueChannel.mockResolvedValue(null);
		await expect(limits.execute(context)).resolves.toBeUndefined();
		// TODO: Somehow assert specifics on this case
	});

	test.each`
		cooldownSeconds | entryDurationMaxSeconds | submissionMaxQuantity
		${null}         | ${null}                 | ${null}
		${42}           | ${null}                 | ${null}
		${null}         | ${42}                   | ${null}
		${null}         | ${null}                 | ${42}
		${42}           | ${42}                   | ${null}
		${null}         | ${42}                   | ${42}
		${42}           | ${null}                 | ${42}
		${42}           | ${42}                   | ${42}
	`(
		"shows statistics on queue limits",
		async ({
			cooldownSeconds,
			entryDurationMaxSeconds,
			submissionMaxQuantity
		}: {
			cooldownSeconds: number | null;
			entryDurationMaxSeconds: number | null;
			submissionMaxQuantity: number | null;
		}) => {
			mockGetStoredQueueConfig.mockResolvedValue({
				cooldownSeconds,
				entryDurationMaxSeconds,
				submissionMaxQuantity,
				blacklistedUsers: [],
				channelId: "",
				queueDurationSeconds: null,
				entryDurationMinSeconds: null
			});
			await expect(limits.execute(context)).resolves.toBeUndefined();
			expect(mockReply).toHaveBeenCalledOnce();

			const replyArgs = mockReply.mock.calls[0] as Array<unknown>;
			expect(replyArgs[0]).toMatchSnapshot();
		}
	);
});
