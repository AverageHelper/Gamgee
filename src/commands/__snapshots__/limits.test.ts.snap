// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Get Queue Limits > shows statistics on queue limits 1`] = `
{
  "embeds": [
    {
      "description": "Use \`?cooldown\` to see your cooldown time",
      "fields": [
        {
          "name": "Number of Submissions:	infinite",
          "value": "The maximum number of submissions that each user may submit.",
        },
        {
          "name": "Submission Cooldown:	none",
          "value": "The minimum amount of time that each user must wait between their own submissions.",
        },
        {
          "name": "Min Song Length:	0",
          "value": "The minimum duration of a song submission.",
        },
        {
          "name": "Max Song Length:	infinite",
          "value": "The maximum duration of a song submission.",
        },
        {
          "name": "Total Queue Length:	infinite",
          "value": "The maximum duration that the queue should take if all its entries were played end-to-end. The queue will automatically close when a submission takes the queue over this limit.",
        },
      ],
      "title": "Queue Limits",
    },
  ],
  "ephemeral": true,
}
`;

exports[`Get Queue Limits > shows statistics on queue limits 2`] = `
{
  "embeds": [
    {
      "description": "Use \`?cooldown\` to see your cooldown time",
      "fields": [
        {
          "name": "Number of Submissions:	infinite",
          "value": "The maximum number of submissions that each user may submit.",
        },
        {
          "name": "Submission Cooldown:	42 seconds",
          "value": "The minimum amount of time that each user must wait between their own submissions.",
        },
        {
          "name": "Min Song Length:	0",
          "value": "The minimum duration of a song submission.",
        },
        {
          "name": "Max Song Length:	infinite",
          "value": "The maximum duration of a song submission.",
        },
        {
          "name": "Total Queue Length:	infinite",
          "value": "The maximum duration that the queue should take if all its entries were played end-to-end. The queue will automatically close when a submission takes the queue over this limit.",
        },
      ],
      "title": "Queue Limits",
    },
  ],
  "ephemeral": true,
}
`;

exports[`Get Queue Limits > shows statistics on queue limits 3`] = `
{
  "embeds": [
    {
      "description": "Use \`?cooldown\` to see your cooldown time",
      "fields": [
        {
          "name": "Number of Submissions:	infinite",
          "value": "The maximum number of submissions that each user may submit.",
        },
        {
          "name": "Submission Cooldown:	none",
          "value": "The minimum amount of time that each user must wait between their own submissions.",
        },
        {
          "name": "Min Song Length:	0",
          "value": "The minimum duration of a song submission.",
        },
        {
          "name": "Max Song Length:	42 seconds",
          "value": "The maximum duration of a song submission.",
        },
        {
          "name": "Total Queue Length:	infinite",
          "value": "The maximum duration that the queue should take if all its entries were played end-to-end. The queue will automatically close when a submission takes the queue over this limit.",
        },
      ],
      "title": "Queue Limits",
    },
  ],
  "ephemeral": true,
}
`;

exports[`Get Queue Limits > shows statistics on queue limits 4`] = `
{
  "embeds": [
    {
      "description": "Use \`?cooldown\` to see your cooldown time",
      "fields": [
        {
          "name": "Number of Submissions:	42",
          "value": "The maximum number of submissions that each user may submit.",
        },
        {
          "name": "Submission Cooldown:	none",
          "value": "The minimum amount of time that each user must wait between their own submissions.",
        },
        {
          "name": "Min Song Length:	0",
          "value": "The minimum duration of a song submission.",
        },
        {
          "name": "Max Song Length:	infinite",
          "value": "The maximum duration of a song submission.",
        },
        {
          "name": "Total Queue Length:	infinite",
          "value": "The maximum duration that the queue should take if all its entries were played end-to-end. The queue will automatically close when a submission takes the queue over this limit.",
        },
      ],
      "title": "Queue Limits",
    },
  ],
  "ephemeral": true,
}
`;

exports[`Get Queue Limits > shows statistics on queue limits 5`] = `
{
  "embeds": [
    {
      "description": "Use \`?cooldown\` to see your cooldown time",
      "fields": [
        {
          "name": "Number of Submissions:	infinite",
          "value": "The maximum number of submissions that each user may submit.",
        },
        {
          "name": "Submission Cooldown:	42 seconds",
          "value": "The minimum amount of time that each user must wait between their own submissions.",
        },
        {
          "name": "Min Song Length:	0",
          "value": "The minimum duration of a song submission.",
        },
        {
          "name": "Max Song Length:	42 seconds",
          "value": "The maximum duration of a song submission.",
        },
        {
          "name": "Total Queue Length:	infinite",
          "value": "The maximum duration that the queue should take if all its entries were played end-to-end. The queue will automatically close when a submission takes the queue over this limit.",
        },
      ],
      "title": "Queue Limits",
    },
  ],
  "ephemeral": true,
}
`;

exports[`Get Queue Limits > shows statistics on queue limits 6`] = `
{
  "embeds": [
    {
      "description": "Use \`?cooldown\` to see your cooldown time",
      "fields": [
        {
          "name": "Number of Submissions:	42",
          "value": "The maximum number of submissions that each user may submit.",
        },
        {
          "name": "Submission Cooldown:	none",
          "value": "The minimum amount of time that each user must wait between their own submissions.",
        },
        {
          "name": "Min Song Length:	0",
          "value": "The minimum duration of a song submission.",
        },
        {
          "name": "Max Song Length:	42 seconds",
          "value": "The maximum duration of a song submission.",
        },
        {
          "name": "Total Queue Length:	infinite",
          "value": "The maximum duration that the queue should take if all its entries were played end-to-end. The queue will automatically close when a submission takes the queue over this limit.",
        },
      ],
      "title": "Queue Limits",
    },
  ],
  "ephemeral": true,
}
`;

exports[`Get Queue Limits > shows statistics on queue limits 7`] = `
{
  "embeds": [
    {
      "description": "Use \`?cooldown\` to see your cooldown time",
      "fields": [
        {
          "name": "Number of Submissions:	42",
          "value": "The maximum number of submissions that each user may submit.",
        },
        {
          "name": "Submission Cooldown:	42 seconds",
          "value": "The minimum amount of time that each user must wait between their own submissions.",
        },
        {
          "name": "Min Song Length:	0",
          "value": "The minimum duration of a song submission.",
        },
        {
          "name": "Max Song Length:	infinite",
          "value": "The maximum duration of a song submission.",
        },
        {
          "name": "Total Queue Length:	infinite",
          "value": "The maximum duration that the queue should take if all its entries were played end-to-end. The queue will automatically close when a submission takes the queue over this limit.",
        },
      ],
      "title": "Queue Limits",
    },
  ],
  "ephemeral": true,
}
`;

exports[`Get Queue Limits > shows statistics on queue limits 8`] = `
{
  "embeds": [
    {
      "description": "Use \`?cooldown\` to see your cooldown time",
      "fields": [
        {
          "name": "Number of Submissions:	42",
          "value": "The maximum number of submissions that each user may submit.",
        },
        {
          "name": "Submission Cooldown:	42 seconds",
          "value": "The minimum amount of time that each user must wait between their own submissions.",
        },
        {
          "name": "Min Song Length:	0",
          "value": "The minimum duration of a song submission.",
        },
        {
          "name": "Max Song Length:	42 seconds",
          "value": "The maximum duration of a song submission.",
        },
        {
          "name": "Total Queue Length:	infinite",
          "value": "The maximum duration that the queue should take if all its entries were played end-to-end. The queue will automatically close when a submission takes the queue over this limit.",
        },
      ],
      "title": "Queue Limits",
    },
  ],
  "ephemeral": true,
}
`;
